dlt/extract/source.py

```diff
@@ -212,6 +212,34 @@ def add_filter(self, item_filter: ItemTransformFunc[bool], insert_at: int = None
             self._pipe.insert_step(FilterItem(item_filter), insert_at)
         return self
 
+    def add_limit(self, max_items: int) -> "DltResource":  # noqa: A003
+        """Adds a limit `max_items` to the resource pipe
+
+        This mutates the encapsulated generator to stop after `max_items` items are yielded. This is useful for testing and debugging. It is
+        a no-op for transformers. Those should be limited by their input data.
+
+        Args:
+            max_items (int): The maximum number of items to yield
+        Returns:
+            "DltResource": returns self
+        """
+        def _gen_wrap(gen: TPipeStep) -> TPipeStep:
+            """Wrap a generator to take the first 50 records"""
+            nonlocal max_items
+            count = 0
+            if inspect.isfunction(gen):
+                gen = gen()
+            for i in gen:  # type: ignore # TODO: help me fix this later
+                yield i
```

Comment:
dlt may close a pipe when another pipe fails. in that case `yield i` raises `GeneratorExit`. in that case you should also close the underlying `gen`

maybe a finally clause would do. also `GeneratorExit` does not derive from Exception but from its base